File: src/java/iced-x86/src/test/java/com/github/icedland/iced/x86/asm/AssemblerLabelTests.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.asm;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import com.github.icedland.iced.x86.CodeWriterImpl;

import static com.github.icedland.iced.x86.asm.AsmRegisters.*;

final class AssemblerLabelTests {
	@Test
	void multiple_labels_on_same_instruction_throws() {
		CodeAssembler c = new CodeAssembler(64);
		CodeLabel l1 = c.createLabel();
		CodeLabel l2 = c.createLabel();
		CodeLabel l3 = c.createLabel();
		c.nop();
		c.label(l1);
		c.nop();
		c.label(l2);
		assertThrows(IllegalArgumentException.class, () -> c.label(l3));
	}

	@Test
	void anonymous_labels() {
		CodeAssembler c = new CodeAssembler(64);

		CodeLabel lbl1 = c.createLabel();
		CodeLabel lbl2 = c.createLabel();
		CodeLabel lbl3 = c.createLabel();
		CodeLabel lbl4 = c.createLabel();

		c.label(lbl1);
		c.inc(eax);
		c.nop();
		c.anonymousLabel();
		c.je(c.b());
		c.nop();
		c.label(lbl2);
		c.je(c.b());
		c.nop();
		c.jmp(lbl1);
		c.nop();
		c.jmp(lbl2);
		c.nop();
		c.jmp(lbl3);
		c.nop();
		c.jmp(lbl4);
		c.nop();
		c.jne(c.f());
		c.nop();
		c.label(lbl3);
		c.jne(c.f());
		c.nop();
		c.anonymousLabel();
		c.inc(eax);
		c.nop();
		c.label(lbl4);
		c.nop();
		c.nop();

		byte[] expectedData = new byte[] {
			(byte)0xFF, (byte)0xC0, (byte)0x90, 0x74, (byte)0xFE, (byte)0x90, 0x74, (byte)0xFB, (byte)0x90, (byte)0xEB, (byte)0xF5, (byte)0x90, (byte)0xEB, (byte)0xF8, (byte)0x90, (byte)0xEB,
			0x07, (byte)0x90, (byte)0xEB, 0x0A, (byte)0x90, 0x75, 0x04, (byte)0x90, 0x75, 0x01, (byte)0x90, (byte)0xFF, (byte)0xC0, (byte)0x90, (byte)0x90, (byte)0x90,
		};
		CodeWriterImpl writer = new CodeWriterImpl();
		c.assemble(writer, 0);
		assertArrayEquals(expectedData, writer.toArray());
	}

	@Test
	void unused_anonymous_label_fails() {
		CodeAssembler c = new CodeAssembler(64);
		c.nop();
		c.anonymousLabel();
		Object result = c.assemble(new CodeWriterImpl(), 0);
		assertInstanceOf(String.class, result);
	}

	@Test
	void undeclared_forward_anon_label_fails() {
		CodeAssembler c = new CodeAssembler(64);
		c.nop();
		c.je(c.f());
		Object result = c.assemble(new CodeWriterImpl(), 0);
		assertInstanceOf(String.class, result);
	}

	@Test
	void at_most_one_anon_label_per_instruction() {
		CodeAssembler c = new CodeAssembler(64);
		c.nop();
		c.anonymousLabel();
		assertThrows(UnsupportedOperationException.class, () -> c.anonymousLabel());
	}

	@Test
	void referencing_backward_anon_label_when_not_defined_throws() {
		CodeAssembler c = new CodeAssembler(64);
		c.nop();
		assertThrows(UnsupportedOperationException.class, () -> c.je(c.b()));
	}

	@Test
	void anonymous_label_and_named_label_cant_use_same_instruction() {
		{
			CodeAssembler c = new CodeAssembler(64);
			CodeLabel lbl = c.createLabel();
			c.nop();
			c.label(lbl);
			c.anonymousLabel();
			assertThrows(UnsupportedOperationException.class, () -> c.nop());
		}
		{
			CodeAssembler c = new CodeAssembler(64);
			CodeLabel lbl = c.createLabel();
			c.nop();
			c.anonymousLabel();
			c.label(lbl);
			assertThrows(UnsupportedOperationException.class, () -> c.nop());
		}
	}
}

File: src/csharp/Intel/Iced.UnitTests/Intel/AssemblerTests/AssemblerLabelTests.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER && CODE_ASSEMBLER
using System;
using Iced.Intel;
using Xunit;
using static Iced.Intel.AssemblerRegisters;

namespace Iced.UnitTests.Intel.AssemblerTests {
	public sealed class AssemblerLabelTests {
		[Fact]
		void Multiple_labels_on_same_instruction_throws() {
			var c = new Assembler(64);
			var l1 = c.CreateLabel();
			var l2 = c.CreateLabel();
			var l3 = c.CreateLabel();
			c.nop();
			c.Label(ref l1);
			c.nop();
			c.Label(ref l2);
			Assert.Throws<ArgumentException>(() => c.Label(ref l3));
		}

		[Fact]
		void Anonymous_labels() {
			var c = new Assembler(64);

			var lbl1 = c.CreateLabel();
			var lbl2 = c.CreateLabel();
			var lbl3 = c.CreateLabel();
			var lbl4 = c.CreateLabel();

			c.Label(ref lbl1);
			c.inc(eax);
			c.nop();
			c.AnonymousLabel();
			c.je(c.@B);
			c.nop();
			c.Label(ref lbl2);
			c.je(c.@B);
			c.nop();
			c.jmp(lbl1);
			c.nop();
			c.jmp(lbl2);
			c.nop();
			c.jmp(lbl3);
			c.nop();
			c.jmp(lbl4);
			c.nop();
			c.jne(c.@F);
			c.nop();
			c.Label(ref lbl3);
			c.jne(c.@F);
			c.nop();
			c.AnonymousLabel();
			c.inc(eax);
			c.nop();
			c.Label(ref lbl4);
			c.nop();
			c.nop();

			var expectedData = new byte[] {
				0xFF, 0xC0, 0x90, 0x74, 0xFE, 0x90, 0x74, 0xFB, 0x90, 0xEB, 0xF5, 0x90, 0xEB, 0xF8, 0x90, 0xEB,
				0x07, 0x90, 0xEB, 0x0A, 0x90, 0x75, 0x04, 0x90, 0x75, 0x01, 0x90, 0xFF, 0xC0, 0x90, 0x90, 0x90,
			};
			var writer = new CodeWriterImpl();
			c.Assemble(writer, 0);
			Assert.Equal(expectedData, writer.ToArray());
		}

		[Fact]
		void Unused_anonymous_label_throws() {
			var c = new Assembler(64);
			c.nop();
			c.AnonymousLabel();
			Assert.Throws<InvalidOperationException>(() => c.Assemble(new CodeWriterImpl(), 0));
		}

		[Fact]
		void Undeclared_forward_anon_label_throws() {
			var c = new Assembler(64);
			c.nop();
			c.je(c.@F);
			Assert.Throws<InvalidOperationException>(() => c.Assemble(new CodeWriterImpl(), 0));
		}

		[Fact]
		void At_most_one_anon_label_per_instruction() {
			var c = new Assembler(64);
			c.nop();
			c.AnonymousLabel();
			Assert.Throws<InvalidOperationException>(() => c.AnonymousLabel());
		}

		[Fact]
		void Referencing_backward_anon_label_when_not_defined_throws() {
			var c = new Assembler(64);
			c.nop();
			Assert.Throws<InvalidOperationException>(() => c.je(c.@B));
		}

		[Fact]
		void Anonymous_label_and_named_label_cant_use_same_instruction() {
			{
				var c = new Assembler(64);
				var lbl = c.CreateLabel();
				c.nop();
				c.Label(ref lbl);
				c.AnonymousLabel();
				Assert.Throws<InvalidOperationException>(() => c.nop());
			}
			{
				var c = new Assembler(64);
				var lbl = c.CreateLabel();
				c.nop();
				c.AnonymousLabel();
				c.Label(ref lbl);
				Assert.Throws<InvalidOperationException>(() => c.nop());
			}
		}
	}
}
#endif
